### DISCUSSION TITLE
Git-based Schema Management w/ Supabase CLI

### DISCUSSION THREAD

**Discussion by devstein** [supabase/supabase#4692] [Feature Requests]
Hi 👋 

First off, want to say thank you for Supabase and all your open source work! I love the way you all have been building the product and keep finding myself thinking how intuitive it is. 

Along those lines, I read [this blog post about migration management with Supabase CLI](https://supabase.com/blog/2021/03/31/supabase-cli#migrations) and have a similar strong preference for automated schema diffing rather than manual migrations. I've been playing around with the Supabase CLI and I would like suggestions on how to achieve the following workflow and if this approach makes sense.

1. The source of truth for the database (non-supabase managed) schemas live in Git: `db/schemas/*`, `db/triggers/*`, etc
2. To make changes to the DB, users modify the `db/**` directory and apply the changes to a fresh local Postgres instance 
     - Essentially running `supabase start` but initializing the schema to the `db/**` directory rather than the state of remote
3. Users use the Supabase CLI to generate the migration, validate it, then push it to the remote 

The main difference is that the source of truth of the state of the database lives in Git and the local development experience is about _declaring_ what you want the new state of the DB to be rather than _modifying_ a local replica (ad-hoc migrations).

**Comment by soedirgo** [comment]
Yup, that's how we wanted it to go at the start. But this turned to be a really hard problem - e.g. you can't differentiate between renaming a column and dropping + creating a column with a different name.

The current CLI still needs the schema diff (so that you can use your own SQL IDE and have the CLI generate new migrations), but it only does so sparingly.

**Reply by devstein** [reply]
@soedirgo Thanks for the quick reply! I'm not sure I understand how the current CLI approach solves this problem. 

For the renaming, a column vs dropping + creating a column example does the current CLI know if a user has run `ALTER TABLE RENAME COLUMN` vs `ALTER TABLE DROP COLUMN`, or do they still have to inspect the migration to verify it's correct?

**Reply by soedirgo** [reply]
Users still have to manually inspect the migrations, but since the migrations is the source of truth rather than a declarative schema, you can do renames.

**Reply by devstein** [reply]
Thanks @soedirgo that makes sense! Appreciate the explanation

**Reply by texastoland** [reply]
I'm currently reading the docs for the first time 🤓

This has been the only place Supabase feels lacking. Otherwise the database CLI and JS API almost completely supersede ORMs like Prisma. EXCEPT for generating migrations from declarative schemas.

I'm considering using Prisma exclusively for this use case like in the following thread:
https://www.reddit.com/r/Supabase/comments/p6mueg/comment/ia10e55/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button
But it can't represent many Postgres features like RLS, materialized views, table inheritance, etc.

> you can't differentiate between renaming a column and dropping + creating a column with a different name

Hypothetically it could be annotated in comments ...

> Users still have to manually inspect the migrations

But that's how Prisma works anyway.

> the migrations is the source of truth rather than a declarative schema

That makes sense ideologically but there's no way to either design/edit (only overwrite) or review/view (because they're cummulative) tables, policies, or stored procedures in code even though they _are_ code.

**Reply by EskelCz** [reply]
@soedirgo Why not save the final schema into an sql file after those migrations? You have it at that point. We would then do the edits like now, generate diff, migrations and get a new "source of truth" sql file we can work with in a text form.

**Comment by jonathan-cfmp** [comment]
@devstein what migration workflow did you end up going with? I am also looking for a declarative schema-based approach as my single source of truth.
